lib/tasks/sync_hosts_with_teamdynamix.rake

```diff
@@ -11,26 +11,77 @@ desc <<-DESC.strip_heredoc.squish
     * rake hosts:sync_with_teamdynamix[hostname1,hostname2,..,hostnameX]
 DESC
 namespace :hosts do
-  task :sync_with_teamdynamix, [:hostnames] => :environment do |_task, args|
-    hostnames = Array.wrap(args.hostnames) + args.extras
-    if hostnames.present?
-      hosts = hostnames.collect { |hostname| Host.find_by(name: hostname) }
-    else
-      hosts = Host.all
-    end
-    error_count = 0
-
-    hosts.each do |host|
-      td_api = host.td_api
-      asset_id = host.teamdynamix_asset_id
-      action = asset_id && td_api.asset_exist?(asset_id) ? :update : :create
-      begin
-        td_api.send("#{action}_asset", host)
-      rescue StandardError => e
-        puts "\n\tS.No | hostname | action | result" if error_count.zero?
-        error_count += 1
-        puts "\t#{error_count} | #{host.name} | #{action} | Failed(#{e.message})"
+  task :sync_with_teamdynamix => :environment do |_task|
+    @td_api = TeamdynamixApi.instance
+    @errors = []
+    @hosts_synced = []
+
+    sync_existing_assets_to_hosts
+
+    create_assets_for_unmapped_hosts
+
+    print_summary
+  end
+
+  # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
+  def sync_existing_assets_to_hosts
+    # sending empty search param to get all the assets in TD at once
+    # Note: due to a limitation(possibly a bug) in TD search asset api
+    #       each and every search payload returns all the assets.
+    @teamdynamix_assets = @td_api.search_asset({})
```

Comment:
Unfortunately, not currently but I think you can open a ticket with TeamDynamix API to get it fixed. 
Once they do, we can pass relevant search param at Line 31. I mean we can actually do that preemptively but it would only work once Teamdymix fixes it. 

What asset attribute would you like to search with? What is the common Asset identifier, we should probably make that an optional configurable item as well.

Reply:
It seems like it wold make our code simpler if we could do:
```ruby
Host.all.each do |h|
  # if there is a TD asset with h.teamdynamix_asset_id update it and next

  # if there is a TD asset with name or serialNumber of host.name then update it and set host.teamdynamix_asset_id and next

  # if there is not an asset then create one and set host.teamdynamix_asset_id
end
```

But if thats really not possible then we have no choice.